perun-wui-registrar/src/main/java/cz/metacentrum/perun/wui/registrar/widgets/items/validators/SshKeysListBoxValidator.java

```diff
@@ -26,34 +32,95 @@ public boolean validateLocal(ListBox listBox) {
 			listBox.setRawStatus(getTransl().cantBeEmpty(), ValidationState.ERROR);
 			return false;
 		}
+		// FIXME - but it is probably not necessary as API should check it
+		for (ExtendedTextBox extendedTextBox : listBox.getListValue()) {
+			String sshKey = extendedTextBox.getValue();
+			if (sshKey.contains(",")) {
+				setResult(Result.INVALID_FORMAT);
+				listBox.setStatus(getTransl().sshKeySeparatorNotAllowed(), ValidationState.ERROR);
+				return false;
+			}
+		}
+		listBox.setStatus(ValidationState.SUCCESS);
+		return true;
+	}
+
+	@Override
+	public void validate(ListBox listBox, Events<Boolean> events) {
+
+		if (!validateLocal(listBox)) {
+			events.onFinished(false);
+			return;
+		}
+
+		if (listBox.getValue() == null || listBox.getValue().isEmpty()) {
+			events.onFinished(true);
+			return;
+		}
+
+		checkVals.clear();
+		indexMap.clear();
 
-		if (listBox.getValue() != null && !listBox.getValue().isEmpty()) {
+		int counter = 1;
+		for (ExtendedTextBox extendedTextBox : listBox.getListValue()) {
+			String sshKey = extendedTextBox.getValue();
+			indexMap.put(sshKey, counter);
+			checkVals.put(sshKey, ValidationState.NONE);
+			counter++;
+		}
 
-			String wrongValues = "";
-			int index = 1;
-			for (ExtendedTextBox extendedTextBox : listBox.getListValue()) {
-				String sshKey = extendedTextBox.getValue();
+		for (String sshKey : checkVals.keySet()) {
 
-				if (sshKey.contains(",")) {
+			UsersManager.validateSSHKey(sshKey, new JsonEvents() {
+				@Override
+				public void onFinished(JavaScriptObject result) {
+					validatingCounter--;
+					checkVals.put(sshKey, ValidationState.SUCCESS);
+					if (validatingCounter == 0) {
+						// last check trigger events
+						for (ValidationState state : checkVals.values()) {
+							// at least one key is invalid -> switch to error
+							if (ValidationState.ERROR == state) {
+								setResult(Result.INVALID_FORMAT);
+								listBox.setRawStatus(getTransl().incorrectFormatItemList() + " <b> <br>" + checkVals.entrySet().stream().filter( entry -> (entry.getValue() == ValidationState.ERROR)).map((entry) -> indexMap.get(entry.getKey()) + ". " +
+										(entry.getKey().length() > 25 ? entry.getKey().substring(0,23) + "..." : entry.getKey())).collect(Collectors.joining("<br>")) + "</b>", ValidationState.ERROR);
+								// pass to outer event as fail
+								events.onFinished(false);
+								return;
+							}
+						}
+						// all values were OK -> trigger success
+						listBox.setStatus(ValidationState.SUCCESS);
+						events.onFinished(true);
+					}
+				}
+
+				@Override
+				public void onError(PerunException error) {
```

Comment:
We should check what Exception was returned from the API, since only `SSHKeyNotValidException` mean that key is invalid. Other errors should be considered as "Can't check key validity". See `UsernameValidator` how it is done.